nucypher/config/keyring.py

```diff
@@ -97,137 +134,207 @@ def _save_public_keyfile(keypath: str, key_data: bytes) -> str:
         keyfile.write(key_data)
-        output_path = keyfile.name
+    return keypath
+
+
+def _write_tls_certificate(certificate: Certificate,
+                           full_filepath: str,
+                           force: bool = False,
+                           ) -> str:
 
-    # TODO: output_path is an integer, who knows why?
-    del keyfile_descriptor
-    return output_path
+    cert_already_exists = os.path.isfile(full_filepath)
+    if force is False and cert_already_exists:
+        raise FileExistsError('A TLS certificate already exists at {}.'.format(full_filepath))
 
+    with open(full_filepath, 'wb') as certificate_file:
+        public_pem_bytes = certificate.public_bytes(TLS_CERTIFICATE_ENCODING)
+        certificate_file.write(public_pem_bytes)
+    return full_filepath
 
-def _derive_key_material_from_passphrase(salt: bytes, passphrase: str) -> bytes:
+
+def _read_tls_public_certificate(filepath: str) -> Certificate:
+    """Deserialize an X509 certificate from a filepath"""
+    try:
+        with open(filepath, 'rb') as certificate_file:
+            cert = x509.load_pem_x509_certificate(certificate_file.read(), backend=default_backend())
+            return cert
+    except FileNotFoundError:
+        raise FileNotFoundError("No SSL certificate found at {}".format(filepath))
+
+
+#
+# Encrypt and Decrypt
+#
+
+def _derive_key_material_from_passphrase(salt: bytes,
+                                         passphrase: str
+                                         ) -> bytes:
     """
     Uses Scrypt derivation to derive a  key for encrypting key material.
     See RFC 7914 for n, r, and p value selections.
     This takes around ~5 seconds to perform.
     """
     key_material = Scrypt(
         salt=salt,
-        length=32,
+        length=__WRAPPING_KEY_LENGTH,
         n=2**20,
         r=8,
         p=1,
         backend=default_backend()
     ).derive(passphrase.encode())
-
     return key_material
 
 
-def _derive_wrapping_key_from_key_material(salt: bytes, key_material: bytes) -> bytes:
+def _derive_wrapping_key_from_key_material(salt: bytes,
+                                           key_material: bytes
+                                           ) -> bytes:
     """
     Uses HKDF to derive a 32 byte wrapping key to encrypt key material with.
     """
     wrapping_key = HKDF(
-        algorithm=hashes.BLAKE2b(64),
-        length=64,
+        algorithm=__HKDF_HASH_ALGORITHM(__HKDF_HASH_LENGTH),
+        length=__HKDF_HASH_LENGTH,
         salt=salt,
-        info=b'NuCypher-KeyWrap',
+        info=__WRAPPING_KEY_INFO,
         backend=default_backend()
     ).derive(key_material)
-
-    return wrapping_key[:32]
+    return wrapping_key[:__WRAPPING_KEY_LENGTH]
 
 
-def _encrypt_umbral_key(wrapping_key: bytes, umbral_key: UmbralPrivateKey) -> dict:
+def _encrypt_umbral_key(wrapping_key: bytes,
+                        umbral_key: UmbralPrivateKey
+                        ) -> Dict[str, bytes]:
```

Comment:
What about using the `UmbralPrivateKey.to_bytes(password)` function?

Reply:
Yes - Good Idea - Made this change in #480; Although, now derivation time for new accounts is > 5 seconds.

Reply:
yeah, you can control that with the `_scrypt_cost` argument. Default value is 20. Anything between 15 and 20 should be fine. Note that the time grows exponentially wrt to this argument. so if you want to make it 4 times faster, a cost of 18 should be OK.

Reply:
Standard cost according to the RFC is 20 for symmetric key derivation.